GrafeasGroup/blossom
# Wiki Subdomain and Additions

## Description:

This change adds a full wiki under `wiki.grafeas.localhost:8000`, which will be able to completely replace the wikijs docker container currently running on the application server. 

## Screenshots:

<img width="1260" alt="Screen Shot 2019-11-24 at 3 59 53 PM" src="https://user-images.githubusercontent.com/5179553/69501617-a6a66600-0ed4-11ea-8b3b-db3ef4d35b79.png">

<img width="1227" alt="Screen Shot 2019-11-24 at 4 00 09 PM" src="https://user-images.githubusercontent.com/5179553/69501619-ab6b1a00-0ed4-11ea-8a37-913d959b07a5.png">

## Testing Instructions:

Launch the server, make migrations and migrate per the instructions in the readme. This change forces the development URL to `grafeas.localhost:8000` instead of just `localhost:8000` -- make sure that you also update your `local_settings.py` file with the version in the README.

If you attempt to load `wiki.grafeas.localhost:8000`, the site should kick you over to the root host's login page: `grafeas.localhost:8000/login/?next=...`. After logging in with the same credentials for the regular site login, it should send you back to the wiki as a logged-in user.

## Other Notes:

There are two security-related hacks that I had to implement in order to get this working. `django-hosts` doesn't allow cycling through all host configurations to find a valid URL, so I had to write that myself in `blossom.authentication.views.LoginView.get_redirect`. Lines 44-49 are a rudimentary domain check; I don't think that it's a security problem as it's only checking if (for example) `wiki.grafeas.org` and `grafeas.org` are the same domain.

The second is in `blossom.wiki.middleware`, which rewrites all calls to the media folder coming from the wiki subdomain to the root domain. The wiki doesn't understand that it's at a subdomain, so it write all attachments and images to disk and stores the url as `wiki.grafeas.localhost:8000/media/wiki/...` instead of `grafeas.localhost...`, which is where the media folder is actually served from. Because the media url calls the subdomain and the subdomain doesn't know what to do with it, it 404s even though the image is on the server. This little middleware fixes that issue and forces the call to load from the right place.

## Checklist:
